server/src/main/java/org/apache/cloudstack/diagnostics/DiagnosticsServiceImpl.java

```diff
@@ -0,0 +1,103 @@
+// under the License.
+package org.apache.cloudstack.diagnostics;
+
+import com.cloud.agent.AgentManager;
+import com.cloud.agent.api.Answer;
+import com.cloud.agent.api.routing.NetworkElementCommand;
+import com.cloud.exception.AgentUnavailableException;
+import com.cloud.exception.InvalidParameterValueException;
+import com.cloud.network.router.RouterControlHelper;
+import com.cloud.utils.component.ManagerBase;
+import com.cloud.utils.component.PluggableService;
+import com.cloud.utils.exception.CloudRuntimeException;
+import com.cloud.vm.NicVO;
+import com.cloud.vm.VMInstanceVO;
+import com.cloud.vm.VirtualMachine;
+import com.cloud.vm.VirtualMachineManager;
+import com.cloud.vm.dao.NicDao;
+import com.cloud.vm.dao.VMInstanceDao;
+import org.apache.cloudstack.api.command.admin.diagnostics.ExecuteDiagnosticsCmd;
+import org.apache.log4j.Logger;
+
+import javax.inject.Inject;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class DiagnosticsServiceImpl extends ManagerBase implements PluggableService, DiagnosticsService {
+    private static final Logger LOGGER = Logger.getLogger(DiagnosticsServiceImpl.class);
+
+    @Inject
+    private RouterControlHelper routerControlHelper;
+    @Inject
+    private AgentManager agentManager;
+    @Inject
+    private VMInstanceDao instanceDao;
+    @Inject
+    private NicDao nicDao;
+    @Inject
+    private VirtualMachineManager vmManager;
+
+    @Override
+    public Map<String, String> runDiagnosticsCommand(final ExecuteDiagnosticsCmd cmd) throws AgentUnavailableException, InvalidParameterValueException {
+        final Long vmId = cmd.getId();
+        final String cmdType = cmd.getType().getValue();
+        final String cmdAddress = cmd.getAddress();
+        final String optionalArguments = cmd.getOptionalArguments();
+        final VMInstanceVO vmInstance = instanceDao.findByIdTypes(vmId, VirtualMachine.Type.ConsoleProxy, VirtualMachine.Type.DomainRouter, VirtualMachine.Type.SecondaryStorageVm);
+
+        if (vmInstance == null) {
+            LOGGER.error("Invalid system vm id provided " + vmId);
+            throw new InvalidParameterValueException("Unable to find a system vm with id " + vmId);
+        }
+        final Long hostId = vmInstance.getHostId();
+
+        if (hostId == null) {
+            LOGGER.error("Unable to find host for virtual machine instance: " + vmInstance.getInstanceName());
+            throw new CloudRuntimeException("Unable to find host for virtual machine instance: " + vmInstance.getInstanceName());
+        }
+
+        final DiagnosticsCommand command = new DiagnosticsCommand(cmdType, cmdAddress, optionalArguments, vmManager.getExecuteInSequence(vmInstance.getHypervisorType()));
+
+        if (vmInstance.getType() == VirtualMachine.Type.DomainRouter) {
+            command.setAccessDetail(NetworkElementCommand.ROUTER_IP, routerControlHelper.getRouterControlIp(vmInstance.getId()));
```

Comment:
Remove the if-else block and replace by something like this:
`networkManager.getSystemVMAccessDetails(vm)`. To use network manager, try something like:
```
    @Inject
    private NetworkOrchestrationService networkManager;
```

Reply:
Resolved, solution implemented and tested on VMware, KVM and Xenserver